src/client.rs

```diff
@@ -280,3 +270,368 @@ where
         Ok(signed)
     }
 }
+
+/// This struct represents a subscription to the progress of some transaction, and is
+/// returned from [`SubmittableExtrinsic::sign_and_submit_then_watch()`].
+pub struct TransactionProgress<'client, T: Config> {
+    sub: Option<RpcSubscription<TransactionStatus<T::Hash, T::Hash>>>,
+    ext_hash: T::Hash,
+    client: &'client Client<T>,
+}
+
+impl<'client, T: Config> std::fmt::Debug for TransactionProgress<'client, T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TransactionProgress")
+            .field("sub", &self.sub)
+            .field("ext_hash", &self.ext_hash)
+            .field("client", &"<client>")
+            .finish()
+    }
+}
+
+impl<'client, T: Config> TransactionProgress<'client, T> {
+    pub(crate) fn new(
+        sub: RpcSubscription<TransactionStatus<T::Hash, T::Hash>>,
+        client: &'client Client<T>,
+        ext_hash: T::Hash,
+    ) -> Self {
+        Self {
+            sub: Some(sub),
+            client,
+            ext_hash,
+        }
+    }
+
+    /// Return the next transaction status when it's emitted.
+    pub async fn next(
+        &mut self,
+    ) -> Result<Option<TransactionProgressStatus<'client, T>>, Error> {
+        // Return `None` if the subscription has been dropped:
+        let sub = match &mut self.sub {
+            Some(sub) => sub,
+            None => return Ok(None),
+        };
+
+        // Return the next item otherwise:
+        let res = sub.next().await?;
+        Ok(res.map(|status| {
+            match status {
+                TransactionStatus::Future => TransactionProgressStatus::Future,
```

Comment:
Would `TransactionStatus::Pending` be more idiomatic?

Reply:
Yep, I reckon so! Buuttt, these variant names are ultimately copied verbatim out of substrate, and need to maintain compatibility with it, so while we _could_ rename the variants without an issue I think, it would probably confuse us when we need to check that they are in sync if we change the names, so I'd be hesitant unless it really helped users (and I reckon we can do that with good docs :)).

Reply:
Fair.

(But yeah, we can obviously open PRs to substrate for things we feel look odd too)